### PR TITLE
scattergl visible restyle fix

### PR DESCRIPTION
**Pull request by etpinard**
fixes https://github.com/plotly/plotly.js/issues/2441 and :palm_tree: a few things in `scattergl/index.js` along the way. 

cc @dfcreative @alexcjohnson

### CODE REVIEW

File: test/jasmine/tests/gl2d_plot_interact_test.js [hunk 1]
```diff
@@ -379,6 +379,48 @@ describe('@gl Test gl2d plots', function() {
...
+            y: [0, 1, -1]
+        }])
+        .then(function() {
+            var scene = gd._fullLayout._plots.xy._scene;
+            spyOn(scene.fill2d, 'draw');
+            spyOn(scene.line2d, 'draw');
+            spyOn(scene.error2d, 'draw');
+            spyOn(scene.scatter2d, 'draw');
+
+            return Plotly.restyle(gd, 'visible', 'legendonly', [0]);
```
Comment (etpinard) [inline test/jasmine/tests/gl2d_plot_interact_test.js]: On `master`, this gives:

peek 2018-03-05 17-29

File: src/traces/scattergl/index.js [hunk 2]
```diff
@@ -556,21 +556,24 @@ function sceneUpdate(gd, subplot) {
         scene.draw = function draw() {
             var i;
             for(i = 0; i < scene.count; i++) {
-                if(scene.fill2d) scene.fill2d.draw(i);
+                if(scene.fill2d && scene.fillOptions[i]) {
+                    // must do all fills first
+                    scene.fill2d.draw(i);
```
Comment (alexcjohnson) [inline src/traces/scattergl/index.js]: Clear improvement over `master` - but this does raise the question of whether this (not your change, but the structure surrounding it) is really the right behavior, particularly with `fill: 'tonext'` but also with multiple fills to zero. I can try this out tomorrow and see if my interpretation of what this is doing is accurate, but In svg `scatter`, each fill gets drawn *below* any trace that it's attached to (even the trace before it if it's a `fill: 'tonext'`) but *above* any other trace that comes before it. If all the lines and points are above all the fills, you can get points and lines that look disconnected from their own fills.

File: src/traces/scattergl/attributes.js [hunk 3]
```diff
@@ -83,9 +83,7 @@ var attrs = module.exports = overrideAll({
         marker: scatterAttrs.unselected.marker
     },
 
-    opacity: extendFlat({}, plotAttrs.opacity, {
-        editType: 'calc'
-    }),
+    opacity: extendFlat({}, plotAttrs.opacity)
```
Comment (alexcjohnson) [inline src/traces/scattergl/attributes.js]: you don't even need the `extendFlat`, right? `overrideAll` will take care of that.

File: src/traces/scattergl/index.js [hunk 4]
```diff
@@ -330,6 +274,33 @@ function sceneOptions(gd, subplot, trace, positions) {
         markerOptions.positions = positions;
     }
 
+    function makeErrorOptions(axLetter, errorOpts, vals) {
```
Comment (dfcreative) [inline src/traces/scattergl/index.js]: Nice!

File: src/traces/scattergl/index.js [hunk 5]
```diff
@@ -486,6 +486,22 @@ function sceneUpdate(gd, subplot) {
     if(!subplot._scene) {
         scene = subplot._scene = Lib.extendFlat({}, reset, first);
 
+        // apply new option to all regl components (used on drag)
+        scene.update = function update(opt) {
```
Comment (etpinard) [inline src/traces/scattergl/index.js]: cc @dfcreative you were right! `scene.update` is still used on drag. Added spy+assertion below to  :lock_with_ink_pen: this thing down.

File: src/traces/scattergl/attributes.js (hunk 3 above)
Comment (etpinard) [reply src/traces/scattergl/attributes.js]: done in https://github.com/plotly/plotly.js/pull/2442/commits/47768321e84cdbd40f4a44bda3e7cd1c0f2c2fae

**Comment by etpinard** [comment]
Dima verbally :white_check_mark: this PR to me.

@alexcjohnson anything else? Your comment in https://github.com/plotly/plotly.js/pull/2442#discussion_r172405040 probably deserved an issue of its own.

**Comment by alexcjohnson** [comment]
💃 
I'll look at the layering issue and make an issue if it's off.

**Pull request merged**